### ISSUE TITLE
Opening a new .aup file with a Macro creates a new unnecessary window (regression)

### ISSUE THREAD

**Issue by Tantacrul** [audacity/audacity#1055]
See my description here:

https://www.dropbox.com/s/zm9yiu1s10ocflw/Open_File_Creates_new_Window.mp4?dl=0

Opening an .aup file in 3.0.2 replaces the initial blank window.

Opening an .aup file in 3.0.3 leaves the initial blank window and creates a new window. This could cause numerous problems with anyone using scripts or macros.

**Comment by petersampsonaudacity** [comment]
This is logged in Bugzilla as P1 Bug #2812
https://bugzilla.audacityteam.org/show_bug.cgi?id=2812
Bug 2812 - Macros: OpenProject2 causes macros to fail

This does NOT happen in 3.0,3 with the File>Open command BUT only in Macros with the  OpenProject2 Macro command

**Comment by Paul-Licameli** [comment]
I'm not sure what to do with the multiplication of ticket systems.  I will write here.

This issue that I opened https://bugzilla.audacityteam.org/show_bug.cgi?id=2764 explains the very good reasons (project corruptions) that some change in behavior of the macro command Open Project2 was needed.  Feel free to reproduce the steps in 3.0.2.

Pull request #909 fixed it.  This was a fix that was careful to change Open Project2 behavior only when really needed to avoid the corruption.  Behavior was not meant to change when importing audio files rather than .aup3.

Now this issue and its duplication at https://bugzilla.audacityteam.org/show_bug.cgi?id=2812 complain that the change in behavior, for .aup3 files (only), breaks certain macros.

But is it consistent with past Audacity releases to elevate that to P1 status as a regression?  It has been the decision of James in the past (though I admit not always agreeing) that we don't guarantee preservation of behavior of macros, or even parsing of them, from release to release.

Yet Peter has also argued for MORE change of Open File 2 behavior, not less, in these issues, in the name of making things more "consistent" :  https://bugzilla.audacityteam.org/show_bug.cgi?id=2798. https://bugzilla.audacityteam.org/show_bug.cgi?id=2797

But it's we who have been inconsistent to date, about what guarantees about macro behavior we mean to give from release to release.  Elevation of this to P1 now isn't well established practice.

(I disagree with that when it comes to the subset of macro functionality formerly called "chains" that apply some effect stack destructively, and useful to VO artists and others as well:  the changes with EQ a few versions back made a lot of them unhappy.  But I agreed or was neutral about the more advanced, and still less popular, use of macros.)

**Comment by petersampsonaudacity** [comment]
> Yet Peter has also argued for MORE change of Open File 2 behavior, not less, in these issues, in the name of making things more "consistent" : https://bugzilla.audacityteam.org/show_bug.cgi?id=2798. https://bugzilla.audacityteam.org/show_bug.cgi?id=2797

What Peter was doing was reporting a regression issue with a Macro that used to work when working with Martin (Tantacrul) on speed testing - an issue reported to me by Martin earlier today.

The issue is NOT that a second window for the opened project is created BUT that it is the original empty window that remains as the active window after the  OpenProject2 command and NOT the newly opened one containing the AUP3 project.  

Thus the Macro then attempts to work on a totally empty project and thereby fails as there is nothing to select or export for the Macro to work on

**Comment by Paul-Licameli** [comment]
I see, let's find an acceptable workaround, even if it means editing those macros.

But the fix for 2764 was surely justified and is not going to be reverted for this.

**Comment by petersampsonaudacity** [comment]
> 
> 
> I see, let's find an acceptable workaround, even if it means editing those macros.
> 
> But the fix for 2764 was surely justified and is not going to be reverted for this.

Yes 2764 fix is fine, I recall testing it and passing it


There is currently no workaround with this logged issue AFAIK to transfer active project status to the project opened with OpenProject2 within the Macro rather than active status remaining with the empty (and thus useless to the Macro) project.

It seems perfectly reasonable to me that Macro user might expect to be able to open a named project and work on it all within the scope of the Macro.

So yes, I agree, lets find a workaround to make this work.

**Comment by Paul-Licameli** [comment]
@Tantacrul was concerned that more than the behavior of macro commands changed between versions.  I say no, the opening and importing commands in the menus have no changes.

Finding a workaround so that macros like @petersampsonaudacity 's for performance testing should be sought, but I say no to a stop-the-press fix for 3.0.3.

The macros you were using were in fact using the Open Project2 command, which I determined to be a dangerous one needing fixing, when I did my own white-box testing of macros, which resulted in several issues in Bugzilla.

The changing of the "active" project during one macro is not supported by the system and won't be easy to do.

The workaround you might think of -- inserting Undo commands into the macro, to put the project back in an empty state, in which it won't open a new window -- is problematic, as I explained in this issue.  https://bugzilla.audacityteam.org/show_bug.cgi?id=2768

In short:  The macro system was not already in my opinion robust enough to carry the weight of a testing suite.  Making it so is worthwhile work, but real enhancement work beyond the scope of a small bug fix for now.

**Comment by petersampsonaudacity** [comment]
It really is a MUSE decision as to what the P level of the bug should be.  I am no longer an "Active" team member having moved recently to "Emeritus" status - so I am no longer the arbiter of QA and in particular Bugzilla.

There are relatively simple workarounds for Tantacrul and his team with that Macro:

a) remove the troublesome OpenProject2 from the Macro,  a project could then be opened manually an the Macro run on that project. 

b) Or, remove the troublesome OpenProject2 from the Macro and add the generation of the necessary audio with a Generator

c) Since it is measuring I/O in various forms add an OpenProject2 to the end of the Macro after the so that Open gets measured there rather than at the beginning.

**Comment by Tantacrul** [comment]
Quick note. It may be our decision but it is still very much the case that we don't always fully appreciate the magnitude of certain bugs because we are still getting to grips with the totality of this complex project. I felt this particular bug was a P2 based on an assumption that it would significantly impact a large (circa 5% or more) number of users.

With that in mind, we do appreciate input on whether new bugs should be promoted or demoted.

**Comment by petersampsonaudacity** [comment]
I suspect that  relatively few IRL users will be using OpenProject2 in a Macro.  Our use here for performance testing is perhaps a special case - but nonetheless a very valid use case.  So P2 would be fine by me as Tantacrul suggests.

a)  There is nothing in the Manual to tell you this is going to happen with OpenProject2

b) there is no error reporting in the Macro to tell you where and why the Macro failed

You and I, Tantacrul, stumbled and fumbled in that Macro this morning at a loss at first to figure out what exactly was going wrong - IRL users would similarly be puzzled and confused.
